Use SQLite key-value store for I18n backend

wjordan opened the pull request:
This PR offloads i18n contents to a SQLite-based key-value storage on disk, in order to reduce memory usage of the running Ruby web-application.

This PR contains 3 high-level components:

1. A new [`i18n`](https://github.com/svenfuchs/i18n) backend `Cdo::I18nBackend < `[`KeyValue`](https://github.com/svenfuchs/i18n/blob/master/lib/i18n/backend/key_value.rb#L25-L69) that automatically synchronizes its contents with Ruby or YAML files in `I18n.load_path`, using the same i18n source files the default `Simple` backend uses.
    - This backend includes the newly-added `CacheFile` module, which stores file-modification times of all loaded Ruby/YAML files so that files are only re-loaded when their modification time changes.
2. A simple key-value class `SqliteKeyValue`, which wraps a `SqLite3::Database` with a simple file-based getter (`[]`) / setter (`[]=`) interface, and a `transaction` method to optimize bulk inserts.
    - SQLite isn't the absolute fastest library for on-disk key-value storage ([`lmdb`](https://symas.com/lmdb/) and [`sparkey`](https://github.com/spotify/sparkey) both performed up to 2x faster in write/read benchmarks I conducted), but we already included SQLite in our stack and it should perform well enough for our purposes, so the choice was the path of least resistance for now.
3. The `KeyValue` store disables 'subtrees' for performance reasons, so a few areas in our application that depended on subtree-keys needed to be rewritten to retrieve full keys directly.

wjordan commented:
Tests are passing, so the next step is profiling on an adhoc instance to compare performance / memory usage against the baseline.

Initial results look promising (~1.15GB memory per Dashboard process versus ~1.8GB baseline, 50sec dashboard-console load versus 1min55sec baseline).

joshlory commented on Gemfile.lock:
```diff
@@ -574,17 +580,17 @@ GEM
+      activejob (= 5.0.7)
+      activemodel (= 5.0.7)
+      activerecord (= 5.0.7)
+      activesupport (= 5.0.7)
```
Intentional to include a Rails patch version upgrade?

wjordan replied:
yeah, in order to upgrade the `i18n` gem to 1.x the Rails patch version needed to be updated (the latest version bumped up the dependency constraint). I may extract the Rails upgrade into a separate PR to merge separately/earlier.

wjordan replied:
extracted the upgrade to #23665

joshlory commented on Gemfile:
```diff
@@ -13,6 +13,8 @@ end
 gem 'rails', '~> 5.0.1'
 gem 'rails-controller-testing'
 
+gem 'i18n', github: 'wjordan/i18n', branch: 'key_value_pluralization_fix'
```
Link to upstream PR: https://github.com/svenfuchs/i18n/pull/419.

wjordan replied:
oh yeah, thanks for the cross-reference. I'll commit an inline comment as well (so we can remember to revert back to upstream if/when the PR merged).

aoby commented on dashboard/app/helpers/notes_helper.rb:
```diff
@@ -3,13 +3,24 @@ module NotesHelper
+      slide_image = try_t("#{slides_key}.#{n}.image", locale: :en)
+      break unless slide_image
+      slides_english[n] = {image: slide_image}
+    end
```
Outside the scope of this PR, but I don't think that comment is accurate. It appears that we populate `slides_english` in all cases, not just for non-english locales.

wjordan replied:
Assuming `I18n.default_locale` is english, the above `return slides` would be invoked if `I18n.locale` is also english, making the comment accurate (since this line is only reached if we're in a non-default [english] locale).

aoby replied:
👍  I think that line was excluded from the original diff. Or maybe I just missed it

aoby commented on dashboard/app/models/levels/blockly.rb:
```diff
@@ -344,9 +344,11 @@ def blockly_level_options
   end
 
-  def get_localized_property(property_name)
+  def get_localized_property(property_name, resolve: true)
```
Why would we not resolve? Please add a small header comment here explaining the params

aoby commented on dashboard/app/models/levels/blockly.rb:
```diff
@@ -344,9 +344,11 @@ def blockly_level_options
+  def get_localized_property(property_name, resolve: true)
     if should_localize? && try(property_name)
-      I18n.t("data.#{property_name.pluralize}.#{name}_#{property_name.singularize}", default: nil)
+      key = "data.#{property_name.pluralize}.#{name}_#{property_name.singularize}"
```
Also outside the scope of this PR, but why `try(property_name)`?

Hamms replied:
Mostly because we use this method to localize properties from `serialized_properties`, which aren't guaranteed to exist

aoby commented on lib/i18n/backend/file_key_value.rb:
```diff
@@ -0,0 +1,75 @@
+    # its contents with Ruby or YAML files on the I18n load path.
+    # This backend improves startup time and reduces memory usage,
+    # and allows you to continue using existing i18n files unchanged.
+    class FileKeyValue < KeyValue
```
This could use some tests

wjordan replied:
agreed- i'll add some coverage before merging.

wjordan replied:
Moved this class into a fork of the `i18n` gem (https://github.com/svenfuchs/i18n/compare/master...wjordan:file_key_value_backend), where I've added some basic test coverage.

Hamms commented on dashboard/app/helpers/notes_helper.rb:
```diff
@@ -3,13 +3,24 @@ module NotesHelper
-    slides = try_t(slides_key)
-    return nil unless slides
+    slides = {}
+    1.step.each do |n|
```
maybe I'm misunderstanding this, but it looks like this (and the slides_english version below) will only grab data from the first slide.

wjordan replied:
[`Numeric#step`](https://ruby-doc.org/core-2.5.0/Numeric.html#method-i-step) invokes the block with a sequence of numbers (incremented by 1 by default), until the `break` is hit.

Hamms commented on dashboard/config/application.rb:
```diff
@@ -80,6 +80,10 @@ class Application < Rails::Application
     # config.i18n.load_path += Dir[Rails.root.join('my', 'locales', '*.{rb,yml}').to_s]
+    require 'i18n/backend/file_key_value'
+    require 'cdo/sqlite_key_value'
+    cache_file = cache_dir('i18n/cached')
```
nit: should `i18n/cached` be a constant somewhere?

Hamms commented on pegasus/src/env.rb:
```diff
@@ -45,15 +48,19 @@ def self.create_logger
   I18n.enforce_available_locales = false
-  I18n.backend.load_translations
+  I18n::Backend::Simple.send(:include, I18n::Backend::Fallbacks)
+  I18n::Backend::FileKeyValue.send(:include, I18n::Backend::Fallbacks)
```
why are we still working on the Simple backend here, rather than just the new one?

wjordan replied:
good point- probably cleaner to call #include on `I18n.backend.class`, so the fallback behavior is included regardless of the underlying backend implementation class.

wjordan replied:
updated in 9e9eb31.

wjordan commented:
Finished my first refactoring pass based on feedback:

- extracted Rails 5.0.7 upgrade to separate PR (#23665)
- extracted FileKeyValue backend to fork of `i18n` gem, filed upstream PR (svenfuchs/i18n#420)
  - included full test coverage in the fork
- Added test coverage for `SqliteKeyValue` implementation
- extracted backend configuration (previously duplicated in dashboard/pegasus) to separate singleton class (`Cdo::I18nBackend`), simplifying the code a bit

Getting close to ready. Here are my Remaining todo items:

- [x] extract code removing subtree usage to separate prerequisite PR (those parts can be merged in separately, before the backend change) - #23745
- [x] add condition allowing backend to be optionally used based on DCDO feature-flag (for initial rollout / testing on canary frontend instance)

wjordan commented:
I think this is ready to merge, as soon as tests pass on the latest rebase.

Since this is change impacts the web-application broadly, I've added the following logic to help with a careful roll-out in production: https://github.com/code-dot-org/code-dot-org/blob/8a371544a2c500db3a3406a2c624c95fd6a70dad/lib/cdo/i18n_backend.rb#L19-L23

This will allow switching between the new and old backends by adjusting the config variable and restarting the web-application services. By default, the new backend will be enabled everywhere except in `production`, where I'd like to manually switch it on selectively (while monitoring resource usage and error-count) before rolling it out entirely.
